plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -606,25 +610,86 @@ public boolean rename(Path src, Path dst)
+        String key = keyFromPath(path);
+        if (recursive) {
+            DeletePrefixResult deletePrefixResult;
+            try {
+                deletePrefixResult = deletePrefix(path);
+            }
+            catch (AmazonClientException e) {
+                throw new IOException("Failed to delete paths with the prefix path " + path, e);
+            }
+            if (deletePrefixResult == DeletePrefixResult.NO_KEYS_FOUND) {
+                deleteObject(key);
+            }
+            else if (deletePrefixResult == DeletePrefixResult.DELETE_KEYS_FAILURE) {
+                return false;
             }
+            deleteObject(key + DIRECTORY_SUFFIX);
         }
-        catch (FileNotFoundException e) {
-            return false;
+        else {
+            if (path.getName().isEmpty()) {
+                return false; // bucket root does not exist
+            }
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(2), true);
+            Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, TrinoS3FileSystem::keysFromRecursiveListing));
+            if (objectKeysIterator.hasNext()) {
+                String directoryKey = objectKeysIterator.next();
+                if (!Objects.equals(directoryKey, key + PATH_SEPARATOR) || objectKeysIterator.hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                deleteObject(directoryKey);
+            }
+            else {
+                deleteObject(key);
+            }
+            deleteObject(key + DIRECTORY_SUFFIX);
         }
+        return true;
+    }
 
-        if (!recursive) {
-            throw new IOException("Directory " + path + " is not empty");
+    private DeletePrefixResult deletePrefix(Path prefix)
+    {
+        String bucketName = getBucketName(uri);
+        Iterator<ListObjectsV2Result> listings = listObjects(prefix, OptionalInt.empty(), true);
+        Iterator<String> objectKeys = Iterators.concat(Iterators.transform(listings, TrinoS3FileSystem::keysFromRecursiveListing));
+        Iterator<List<String>> objectKeysBatches = Iterators.partition(objectKeys, BATCH_DELETE_OBJECTS_REQEST_MAX_PAGE_SIZE);
+        if (!objectKeysBatches.hasNext()) {
+            return DeletePrefixResult.NO_KEYS_FOUND;
         }
 
-        for (FileStatus file : listStatus(path)) {
-            delete(file.getPath(), true);
+        boolean allKeysDeleted = true;
+        while (objectKeysBatches.hasNext()) {
+            String[] objectKeysBatch = objectKeysBatches.next().toArray(String[]::new);
+            try {
+                DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucketName).withKeys(objectKeysBatch);
+                if (requesterPaysEnabled) {
+                    deleteObjectsRequest.setRequesterPays(true);
+                }
+                s3.deleteObjects(deleteObjectsRequest);
```

Comment:
As per `DeleteObjectsResult` javadoc this object is being returned in case that all keys submitted for deletion where successfully deleted. 

https://github.com/aws/aws-sdk-java/blob/8f6c93e49024049d2e5d882a3e2cd0e99e55ee65/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/DeleteObjectsResult.java#L23-L30